### DISCUSSION TITLE
CQL Approach for FHIR Core

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#611] [Q&A]
[**CQL**](https://cql.hl7.org/) (Clinical Quality Language) is the HL7 standard for providing logic execution for decision support and clinical quality measures. CQL allows for a more modular, flexible, and robust expression of logic. It allows logic to be shared between measures and with clinical decision support. CQL provides the ability to express logic that is human-readable yet structured enough for processing a query electronically. The CQL standards are located on the HL7 Website: CQL [Specification](http://www.hl7.org/implement/standards/product_brief.cfm?product_id=400). CQL Examples can be found [here](http://build.fhir.org/ig/HL7/cql/examples.html)

Our initial goal is to design and implement a  CQL Library to demo the below functionality 
1. Generate a MeasureReport via $measure operation 
2. Load and execute a custom CQL library for decision logic
3. Transform a PlanDefinition to a CarePlan using $apply operation

**Main use cases to model** 
1. How to programmatically implement $apply operation i.e. Use the **$apply** to generate a CarePlan from a PlanDefinition
2. Generate a MeasureReport by running the **$evaluate-measure**  tracked here #202

**Related subtasks** 
- [ ] How to run $apply on Android #309
- [x] How to run $measure-evaluate on Android #310
- [x] Integrate CQL Library demo app into FHIR core #177
- [ ] Patient CarePlan generation / $apply operation #100
- [x] Create a Plan Definition for the Covax App #101
- [ ] Create a PlanDefiniton for the CHW/ANC app -  #283

**Comment by MuhammadSalman-7214** [comment]
After reading the documentation of CQL and exploring CQL, here are some open questions for further research and discussion :

1. What will be the conditions to generate a CQL library for Care Plan / $Apply Operation?
2.  What will be the conditions to generate CQL for the measure report?
3.  How will we sync it to the server?
4.  Where will we write the CQL library(CQL expressions that will be created based on our conditions)?

**Reply by kimigxfoxy** [reply]
So the current Measure Reporting and CQL in FHIRCore uses the https://github.com/DBCG/cql_engine dependency. This implementation is a POC. The usage of the above dependency is as follows:

1. Add the measure reporting and CQL dependencies to a project gradle file (https://github.com/DBCG/CqlEvaluatorSampleApp/blob/master/app/build.gradle).
2. For CQL, the following data needs to be loaded first (also shown in an example in the dependency mentioned); 
CQL main FHIR library (https://fhir.labs.smartregister.org/fhir/Library?_id=1752), 
CQL FHIR helper libraries (https://fhir.labs.smartregister.org/fhirLibrary?_id=1753), 
CQL FHIR ValueSet (https://fhir.labs.smartregister.org/fhir/ValueSet?_id=1750,1751)
and a FHIR patient to apply the operation on.  All these data is loaded from the server. After the data is loaded from the server the CQL apply operation is run (Also shown in an example in the dependency mentioned). This returns a JSON string with values in a boolean format.

`{
   "resourceType":"Parameters",
   "parameter":[
      {
         "name":"Patient"
      },
      {
         "name":"Estimated Due Date"
      },
      {
         "name":"Gestational Age in Weeks"
      },
      {
         "name":"Hb Concentration"
      },
      {
         "name":"Has Anaemia"
      },
      {
         "name":"Pregnancy Status"
      },
      {
         "name":"Pregnant",
         "valueBoolean":false
      },
      {
         "name":"Is Recommendation Applicable",
         "valueBoolean":false
      },
      {
         "name":"Get Card Summary",
         "valueString":"Recommend 30-60 mg elemental iron and 0.4 mg folic acid, daily"
      },
      {
         "name":"Get Card Detail",
         "valueString":"Daily elemental iron of between 30m and 60mg, and daily dose of 400 ug (0.4 mg) of folic acid is recommended for pregnant women"
      },
      {
         "name":"Get Card Indicator",
         "valueString":"info"
      }
   ]
}`

3. For measure reporting,  the following data needs to be loaded first (also shown in an example in the dependency mentioned);  FHIR measure reporting libraries (https://hapi.fhir.org/baseR4/Library?_id=ANCDataElements,WHOCommon,ANCConcepts,ANCContactDataElements,FHIRHelpers,ANCStratifiers,ANCIND01,ANCCommon,ANCBaseDataElements,FHIRCommon,ANCBaseConcepts: https://hapi.fhir.org/baseR4/Measure?_id=ANCIND01
), 
FHIR patients to apply the operation on, report type (Individual or Population), report name (e.g. Mothers who missed first ANC visit), date period from and date period to. The libraries are also loaded from the server. After the data is loaded from the server the measure report apply operation is run (Also shown in an example in the dependency mentioned). This returns a JSON string with values in a boolean format,  a population numerator and a population denominator.

Using the https://github.com/DBCG/cql_engine dependency engine provided the following challenges.

1. The POC examples are not well written to optimize memory and occasionally there were unnecessary objects created at runtime which will cause out of memory exceptions. A fix for this is to rewrite the load and apply operations in FHIRCore to fix memory leaks during object creations.
2. The requirement in FHIRCore is to load FHIR libraries from the server so as to apply the operations offline. This posed a challenge since many of the libraries were huge and caused out of memory exceptions. In the case of measure reporting, the FHIR libraries to be loaded were of 52 MB in size, nearly 3 times the size of a standard android application. This caused a lot of out of memory exceptions and a work around for this was to load the libraries in chunks using many REST request to the server and finally concatenating them together in a variable. This variable, will still however hold the application at ransom since it was occupying 52MBs and the only way to get back that memory was to wait for the android Garbage Collector to clear it once it was no longer in use.
3. Since the libraries loading were taking a lot of time to load and they were not changing after a longtime , another work around was to load the libraries from the server only once and for the first time when the user opens the reporting module and writing these libraries in the internal memory of the application. The next time the user opens the reporting module these libraries will be loaded first and will not have to wait for over 30 seconds to get the data from the server. This solved the load time issue but not the out of memory issue.
4. Also noted while implementing measure reporting, **_the results gotten from the engine were not consistent_**. For example. the population numerators and denominators were not changing thus making the report questionable.
5. Also saving the libraries into our local HAPI FHIR server (https://fhir.labs.smartregister.org/fhir/) was not easy. The server was not accepting the huge 52mbs of libraries. The work around was to use a public HAPI FHIR server (https://hapi.fhir.org/baseR4/) which fortunately accepted the data.

Currently issue https://github.com/google/android-fhir/issues/657 and PR https://github.com/google/android-fhir/pull/810 are being implemented in android FHIR. The hope is it may fix memory issues, operation times and results credibility. More importantly is to understand how it will solve out of memory issues when huge loads of FHIR resources are loaded, will a need to do it on server be considered? Or is their a way to use smaller libraries

**Reply by pld** [reply]
Do you have examples of the engine being inconsistent? Can we write up a test case for this and open an issue against the CQL repo about it?

This is likely to be a problem for a lot of groups.
